pkg/download/download.go

```diff
@@ -0,0 +1,168 @@
+		}
+		return filterAPIList, nil
+	}
+	requestedApis := strings.Split(downloadSpecificAPI, ",")
+	result := true
+	errString := ""
+
+	for _, id := range requestedApis {
+		cleanAPI := strings.TrimSpace(id)
+		isApi := api.IsApi(cleanAPI)
+		if isApi == false {
+			result = false
+			errString += errString + fmt.Sprintf(" \t - Value %s is not a valid API name.\n", cleanAPI)
+		} else {
+			filterAPI := availableApis[cleanAPI]
+			path := transFormSpecialCasesAPIPath(filterAPI.GetId(), filterAPI.GetApiPath())
+			filterAPIList[cleanAPI] = api.NewApi(filterAPI.GetId(), path)
+		}
+	}
+	if result == false {
+		return nil, errors.New(errString)
+	}
+	return filterAPIList, nil
+}
+
+//function that deals with modifying the api path register in the api class to apply filters to skip read only entities from being downloaded
+func transFormSpecialCasesAPIPath(apiID string, apiURL string) string {
+
+	switch apiID {
+	case "synthetic-location":
+		return apiURL + "?type=PRIVATE"
+	default:
+		return apiURL
+	}
+}
+
+//creates the project and dowload the configs
+func downloadConfigFromEnvironment(environment environment.Environment, basepath string, listApis map[string]api.Api) (err error) {
+
+	projectName := baseProjectName + environment.GetId()
+	path := filepath.Join(basepath, projectName)
+	creator := files.NewDiskFileCreator()
+
+	util.Log.Info("Creating base project name %s", projectName)
+	fullpath, err := creator.CreateFolder(path)
+	if err != nil {
+		return err
+	}
+	token, err := environment.GetToken()
+	if err != nil {
+		util.Log.Error("error retrieving token: %s %v", err)
+		return err
+	}
+	client, err := rest.NewDynatraceClient(environment.GetEnvironmentUrl(), token)
+	if err != nil {
+		util.Log.Error("error creating dynatrace client: %s %v", err)
+		return err
+	}
+	for _, api := range listApis {
+		util.Log.Info(" --- GETTING CONFIGS for %s", api.GetId())
+		jcreator := jsoncreator.NewJSONCreator()
+		ycreator := yamlcreator.NewYamlConfig()
+		err = createConfigsFromAPI(api, token, creator, fullpath, client, jcreator, ycreator)
+		if err != nil {
+			util.Log.Error("error configs for api: %s %v", api.GetId(), err)
+		}
+	}
+	util.Log.Info("END downloading info %s", projectName)
+	return err
+}
+
+func createConfigsFromAPI(api api.Api, token string, creator files.FileCreator, fullpath string, client rest.DynatraceClient,
+	jcreator jsoncreator.JSONCreator, ycreator yamlcreator.YamlCreator) (err error) {
+	//retrieves all objects for the specific api
+	values, err := client.List(api)
+
+	if len(values) == 0 {
+		util.Log.Info("No elements for API %s", api.GetId())
+		return nil
+	}
+	subPath := filepath.Join(fullpath, api.GetId())
+	creator.CreateFolder(subPath)
+
+	for _, val := range values {
+		util.Log.Debug("getting detail %s", val)
+		cont++
+		util.Log.Debug("REQUEST counter %v", cont)
+		name, filter, err := jcreator.CreateJSONConfig(client, api, val, creator, subPath)
+		if err != nil {
```

Comment:
Does it make sense to collect these errors and return them?

Reply:
yes I don't think that the application should stop the download process if only one of the API is failing

Reply:
I agree with @diegorqc , as long as we ensure that all errors are logged in a usable manner at the end this is similar to how we handle upload.
Just cause one environment fails to download a single object, it's more user friendly (and realistic) to continue on for the X other configs and environments

Reply:
With that I would consider collecting and adding these errors to the one's you log with `publishErrors` though rather than logging them here